File: Samples for Solid Edge/[VB]-Samples for Solid Edge/VB/SolidEdge.AddIn.Basic/MyAddIn.vb
Imports Microsoft.Win32
Imports System.Runtime.InteropServices
Imports System.Runtime.InteropServices.ComTypes
Imports System.Text

Namespace SolidEdge.AddIn.Basic
    <Guid("5F47DA3A-B1CF-4B80-9038-C7823CD102E6"), ProgId("SolidEdge.AddIn.Basic"), ComVisible(True), AddInInfo("CodePlex Basic Addin", "Solid Edge Addin in .NET 4.0."), AddInEnvironmentCategory(CategoryIDs.CATID_SEApplication), AddInEnvironmentCategory(CategoryIDs.CATID_SEAllDocumentEnvrionments), AddInAutoConnect(True)> _
    Public Class MyAddIn
        Implements SolidEdgeFramework.ISolidEdgeAddIn, SolidEdgeFramework.ISEAddInEvents, SolidEdgeFramework.ISEApplicationEvents

        Private _application As SolidEdgeFramework.Application
        Private _addInEx As SolidEdgeFramework.ISEAddInEx
        Private _connectionPointDictionary As New Dictionary(Of IConnectionPoint, Integer)()

#Region "SolidEdgeFramework.ISolidEdgeAddIn"

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISolidEdgeAddIn.OnConnection().
        ''' </summary>
        Public Sub OnConnection(ByVal Application As Object, ByVal ConnectMode As SolidEdgeFramework.SeConnectMode, ByVal AddInInstance As SolidEdgeFramework.AddIn) Implements SolidEdgeFramework.ISolidEdgeAddIn.OnConnection
            _application = CType(Application, SolidEdgeFramework.Application)
            _addInEx = CType(AddInInstance, SolidEdgeFramework.ISEAddInEx)

            HookEvents(_addInEx, GetType(SolidEdgeFramework.ISEAddInEvents).GUID)
            HookEvents(_application, GetType(SolidEdgeFramework.ISEApplicationEvents).GUID)

            Select Case ConnectMode
                Case SolidEdgeFramework.SeConnectMode.seConnectAtStartup
                Case SolidEdgeFramework.SeConnectMode.seConnectByUser
                Case SolidEdgeFramework.SeConnectMode.seConnectExternally
            End Select
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISolidEdgeAddIn.OnConnectToEnvironment().
        ''' </summary>
        Public Sub OnConnectToEnvironment(ByVal EnvCatID As String, ByVal pEnvironmentDispatch As Object, ByVal bFirstTime As Boolean) Implements SolidEdgeFramework.ISolidEdgeAddIn.OnConnectToEnvironment
            Dim envGuid As New Guid(EnvCatID)
            Dim environment As SolidEdgeFramework.Environment = CType(pEnvironmentDispatch, SolidEdgeFramework.Environment)

            ' Demonstrate working with CategoryIDs.
            If envGuid.Equals(CategoryIDs.CATID_SEPart) Then
            ElseIf envGuid.Equals(CategoryIDs.CATID_SEDMPart) Then
            End If

            ' Some things only need to be done during bFirstTime.
            If bFirstTime Then
            End If

        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISolidEdgeAddIn.OnDisconnection().
        ''' </summary>
        Public Sub OnDisconnection(ByVal DisconnectMode As SolidEdgeFramework.SeDisconnectMode) Implements SolidEdgeFramework.ISolidEdgeAddIn.OnDisconnection
            Select Case DisconnectMode
                Case SolidEdgeFramework.SeDisconnectMode.seDisconnectAtShutdown
                Case SolidEdgeFramework.SeDisconnectMode.seDisconnectByUser
                Case SolidEdgeFramework.SeDisconnectMode.seDisconnectExternally
            End Select

            ' Unhook all events.
            UnhookAllEvents()
        End Sub

#End Region

#Region "SolidEdgeFramework.ISEAddInEvents"

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEAddInEvents.OnCommand event.
        ''' </summary>
        Public Sub OnCommand(ByVal CommandID As Integer) Implements SolidEdgeFramework.ISEAddInEvents.OnCommand
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEAddInEvents.OnCommandHelp event.
        ''' </summary>
        Public Sub OnCommandHelp(ByVal hFrameWnd As Integer, ByVal HelpCommandID As Integer, ByVal CommandID As Integer) Implements SolidEdgeFramework.ISEAddInEvents.OnCommandHelp
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEAddInEvents.OnCommandUpdateUI event.
        ''' </summary>
        Public Sub OnCommandUpdateUI(ByVal CommandID As Integer, ByRef CommandFlags As Integer, <System.Runtime.InteropServices.Out()> ByRef MenuItemText As String, ByRef BitmapID As Integer) Implements SolidEdgeFramework.ISEAddInEvents.OnCommandUpdateUI
            MenuItemText = String.Empty
        End Sub

#End Region

#Region "SolidEdgeFramework.ISEApplicationEvents"

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterActiveDocumentChange event.
        ''' </summary>
        Public Sub AfterActiveDocumentChange(ByVal theDocument As Object) Implements SolidEdgeFramework.ISEApplicationEvents.AfterActiveDocumentChange
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterCommandRun event.
        ''' </summary>
        Public Sub AfterCommandRun(ByVal theCommandID As Integer) Implements SolidEdgeFramework.ISEApplicationEvents.AfterCommandRun
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterDocumentOpen event.
        ''' </summary>
        Public Sub AfterDocumentOpen(ByVal theDocument As Object) Implements SolidEdgeFramework.ISEApplicationEvents.AfterDocumentOpen
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterDocumentPrint event.
        ''' </summary>
        Public Sub AfterDocumentPrint(ByVal theDocument As Object, ByVal hDC As Integer, ByRef ModelToDC As Double, ByRef Rect As Integer) Implements SolidEdgeFramework.ISEApplicationEvents.AfterDocumentPrint
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterDocumentSave event.
        ''' </summary>
        Public Sub AfterDocumentSave(ByVal theDocument As Object) Implements SolidEdgeFramework.ISEApplicationEvents.AfterDocumentSave
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterEnvironmentActivate event.
        ''' </summary>
        Public Sub AfterEnvironmentActivate(ByVal theEnvironment As Object) Implements SolidEdgeFramework.ISEApplicationEvents.AfterEnvironmentActivate
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterNewDocumentOpen event.
        ''' </summary>
        Public Sub AfterNewDocumentOpen(ByVal theDocument As Object) Implements SolidEdgeFramework.ISEApplicationEvents.AfterNewDocumentOpen
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterNewWindow event.
        ''' </summary>
        Public Sub AfterNewWindow(ByVal theWindow As Object) Implements SolidEdgeFramework.ISEApplicationEvents.AfterNewWindow
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterWindowActivate event.
        ''' </summary>
        Public Sub AfterWindowActivate(ByVal theWindow As Object) Implements SolidEdgeFramework.ISEApplicationEvents.AfterWindowActivate
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.BeforeCommandRun event.
        ''' </summary>
        Public Sub BeforeCommandRun(ByVal theCommandID As Integer) Implements SolidEdgeFramework.ISEApplicationEvents.BeforeCommandRun
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.BeforeDocumentClose event.
        ''' </summary>
        Public Sub BeforeDocumentClose(ByVal theDocument As Object) Implements SolidEdgeFramework.ISEApplicationEvents.BeforeDocumentClose
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.BeforeDocumentPrint event.
        ''' </summary>
        Public Sub BeforeDocumentPrint(ByVal theDocument As Object, ByVal hDC As Integer, ByRef ModelToDC As Double, ByRef Rect As Integer) Implements SolidEdgeFramework.ISEApplicationEvents.BeforeDocumentPrint
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.BeforeDocumentSave event.
        ''' </summary>
        Public Sub BeforeDocumentSave(ByVal theDocument As Object) Implements SolidEdgeFramework.ISEApplicationEvents.BeforeDocumentSave
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.BeforeEnvironmentDeactivate event.
        ''' </summary>
        Public Sub BeforeEnvironmentDeactivate(ByVal theEnvironment As Object) Implements SolidEdgeFramework.ISEApplicationEvents.BeforeEnvironmentDeactivate
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.BeforeQuit event.
        ''' </summary>
        Public Sub BeforeQuit() Implements SolidEdgeFramework.ISEApplicationEvents.BeforeQuit
        End Sub

        ''' <summary>
        ''' Implementation of SolidEdgeFramework.ISEApplicationEvents.BeforeWindowDeactivate event.
        ''' </summary>
        Public Sub BeforeWindowDeactivate(ByVal theWindow As Object) Implements SolidEdgeFramework.ISEApplicationEvents.BeforeWindowDeactivate
        End Sub

#End Region

#Region "IConnectionPoint helpers"

        ''' <summary>
        ''' Attaches specified events to this object.
        ''' </summary>
        Private Sub HookEvents(ByVal eventSource As Object, ByVal eventGuid As Guid)
            Dim container As IConnectionPointContainer = Nothing
            Dim connectionPoint As IConnectionPoint = Nothing
            Dim cookie As Integer = 0

            container = CType(eventSource, IConnectionPointContainer)
            container.FindConnectionPoint(eventGuid, connectionPoint)

            If connectionPoint IsNot Nothing Then
                connectionPoint.Advise(Me, cookie)
                _connectionPointDictionary.Add(connectionPoint, cookie)
            End If
        End Sub

        ''' <summary>
        ''' Detaches specified events from this object.
        ''' </summary>
        Private Sub UnhookAllEvents()
            Dim enumerator As Dictionary(Of IConnectionPoint, Integer).Enumerator = _connectionPointDictionary.GetEnumerator()
            Do While enumerator.MoveNext()
                enumerator.Current.Key.Unadvise(enumerator.Current.Value)
            Loop

            _connectionPointDictionary.Clear()
        End Sub

#End Region

#Region "regasm.exe"

        ''' <summary>
        ''' Implementation of ComRegisterFunction.
        ''' </summary>
        ''' <remarks>
        ''' This method gets called when regasm.exe is executed against the assembly.
        ''' </remarks>
        <ComRegisterFunction> _
        Public Shared Sub Register(ByVal t As Type)
            Dim info = CType(AddInInfoAttribute.GetCustomAttribute(t, GetType(AddInInfoAttribute)), AddInInfoAttribute)
            Dim environments = CType(AddInEnvironmentCategoryAttribute.GetCustomAttributes(t, GetType(AddInEnvironmentCategoryAttribute)), AddInEnvironmentCategoryAttribute())
            Dim autoConnect = CType(AddInAutoConnectAttribute.GetCustomAttribute(t, GetType(AddInAutoConnectAttribute)), AddInAutoConnectAttribute)

            If info Is Nothing Then
                Throw New System.Exception("Missing AddInInfoAttribute.")
            End If
            If (environments Is Nothing) OrElse (environments.Length = 0) Then
                Throw New System.Exception("Missing AddInEnvironmentCategoryAttribute.")
            End If
            If autoConnect Is Nothing Then
                autoConnect = New AddInAutoConnectAttribute(True)
            End If

            Dim subkey As String = String.Format("CLSID\{0}", t.GUID.ToString("B"))
            Using baseKey As RegistryKey = Registry.ClassesRoot.CreateSubKey(subkey)
                subkey = String.Format("Implemented Categories\{0}", CategoryIDs.CATID_SolidEdgeAddIn)
                Using implementedCategoriesKey As RegistryKey = baseKey.CreateSubKey(subkey)
                End Using

                For Each environment In environments
                    subkey = String.Format("Environment Categories\{0}", environment.Guid.ToString("B"))
                    Using environmentCategoryKey As RegistryKey = baseKey.CreateSubKey(subkey)
                    End Using
                Next environment

                Using summaryKey As RegistryKey = baseKey.CreateSubKey("Summary")
                    summaryKey.SetValue("409", info.Summary)
                End Using

                baseKey.SetValue("AutoConnect", If(autoConnect.AutoConnect, 1, 0))
                baseKey.SetValue("409", info.Title)
            End Using
        End Sub

        ''' <summary>
        ''' Implementation of ComUnregisterFunction.
        ''' </summary>
        ''' <remarks>
        ''' This method gets called when regasm.exe is executed against the assembly.
        ''' </remarks>
        <ComUnregisterFunction> _
        Public Shared Sub Unregister(ByVal t As Type)
            Dim subkey As String = String.Format("CLSID\{0}", t.GUID.ToString("B"))
            Registry.ClassesRoot.DeleteSubKeyTree(subkey, False)
        End Sub

#End Region

#Region "Properties"

        Public ReadOnly Property Application() As SolidEdgeFramework.Application
            Get
                Return _application
            End Get
        End Property
        Public ReadOnly Property AddIn() As SolidEdgeFramework.ISEAddInEx
            Get
                Return _addInEx
            End Get
        End Property

#End Region
    End Class
End Namespace

File: Samples for Solid Edge/[C#]-Samples for Solid Edge/C#/SolidEdge.AddIn.Basic/MyAddIn.cs
using Microsoft.Win32;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Runtime.InteropServices.ComTypes;
using System.Text;

namespace SolidEdge.AddIn.Basic
{
    [Guid("5F47DA3A-B1CF-4B80-9038-C7823CD102E6")] // Must be unique!
    [ProgId("SolidEdge.AddIn.Basic")] // Must be unique!
    [ComVisible(true)]
    [AddInInfo("CodePlex Basic Addin", "Solid Edge Addin in .NET 4.0.")]
    [AddInEnvironmentCategory(CategoryIDs.CATID_SEApplication)]
    [AddInEnvironmentCategory(CategoryIDs.CATID_SEAllDocumentEnvrionments)]
    [AddInAutoConnect(true)]
    public class MyAddIn
        : SolidEdgeFramework.ISolidEdgeAddIn, SolidEdgeFramework.ISEAddInEvents, SolidEdgeFramework.ISEApplicationEvents
    {
        private SolidEdgeFramework.Application _application;
        private SolidEdgeFramework.ISEAddInEx _addInEx;
        private Dictionary<IConnectionPoint, int> _connectionPointDictionary = new Dictionary<IConnectionPoint, int>();

        #region SolidEdgeFramework.ISolidEdgeAddIn

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISolidEdgeAddIn.OnConnection().
        /// </summary>
        public void OnConnection(object Application, SolidEdgeFramework.SeConnectMode ConnectMode, SolidEdgeFramework.AddIn AddInInstance)
        {
            _application = (SolidEdgeFramework.Application)Application;
            _addInEx = (SolidEdgeFramework.ISEAddInEx)AddInInstance;

            HookEvents(_addInEx, typeof(SolidEdgeFramework.ISEAddInEvents).GUID);
            HookEvents(_application, typeof(SolidEdgeFramework.ISEApplicationEvents).GUID);

            switch (ConnectMode)
            {
                case SolidEdgeFramework.SeConnectMode.seConnectAtStartup:
                    break;
                case SolidEdgeFramework.SeConnectMode.seConnectByUser:
                    break;
                case SolidEdgeFramework.SeConnectMode.seConnectExternally:
                    break;
            }
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISolidEdgeAddIn.OnConnectToEnvironment().
        /// </summary>
        public void OnConnectToEnvironment(string EnvCatID, object pEnvironmentDispatch, bool bFirstTime)
        {
            Guid envGuid = new Guid(EnvCatID);
            SolidEdgeFramework.Environment environment = (SolidEdgeFramework.Environment)pEnvironmentDispatch;

            // Demonstrate working with CategoryIDs.
            if (envGuid.Equals(CategoryIDs.CATID_SEPart))
            {
            }
            else if (envGuid.Equals(CategoryIDs.CATID_SEDMPart))
            {
            }

            // Some things only need to be done during bFirstTime.
            if (bFirstTime)
            {
            }

        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISolidEdgeAddIn.OnDisconnection().
        /// </summary>
        public void OnDisconnection(SolidEdgeFramework.SeDisconnectMode DisconnectMode)
        {
            switch (DisconnectMode)
            {
                case SolidEdgeFramework.SeDisconnectMode.seDisconnectAtShutdown:
                    break;
                case SolidEdgeFramework.SeDisconnectMode.seDisconnectByUser:
                    break;
                case SolidEdgeFramework.SeDisconnectMode.seDisconnectExternally:
                    break;
            }

            // Unhook all events.
            UnhookAllEvents();
        }

        #endregion

        #region SolidEdgeFramework.ISEAddInEvents

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEAddInEvents.OnCommand event.
        /// </summary>
        public void OnCommand(int CommandID)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEAddInEvents.OnCommandHelp event.
        /// </summary>
        public void OnCommandHelp(int hFrameWnd, int HelpCommandID, int CommandID)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEAddInEvents.OnCommandUpdateUI event.
        /// </summary>
        public void OnCommandUpdateUI(int CommandID, ref int CommandFlags, out string MenuItemText, ref int BitmapID)
        {
            MenuItemText = String.Empty;
        }

        #endregion

        #region SolidEdgeFramework.ISEApplicationEvents

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterActiveDocumentChange event.
        /// </summary>
        public void AfterActiveDocumentChange(object theDocument)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterCommandRun event.
        /// </summary>
        public void AfterCommandRun(int theCommandID)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterDocumentOpen event.
        /// </summary>
        public void AfterDocumentOpen(object theDocument)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterDocumentPrint event.
        /// </summary>
        public void AfterDocumentPrint(object theDocument, int hDC, ref double ModelToDC, ref int Rect)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterDocumentSave event.
        /// </summary>
        public void AfterDocumentSave(object theDocument)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterEnvironmentActivate event.
        /// </summary>
        public void AfterEnvironmentActivate(object theEnvironment)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterNewDocumentOpen event.
        /// </summary>
        public void AfterNewDocumentOpen(object theDocument)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterNewWindow event.
        /// </summary>
        public void AfterNewWindow(object theWindow)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.AfterWindowActivate event.
        /// </summary>
        public void AfterWindowActivate(object theWindow)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.BeforeCommandRun event.
        /// </summary>
        public void BeforeCommandRun(int theCommandID)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.BeforeDocumentClose event.
        /// </summary>
        public void BeforeDocumentClose(object theDocument)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.BeforeDocumentPrint event.
        /// </summary>
        public void BeforeDocumentPrint(object theDocument, int hDC, ref double ModelToDC, ref int Rect)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.BeforeDocumentSave event.
        /// </summary>
        public void BeforeDocumentSave(object theDocument)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.BeforeEnvironmentDeactivate event.
        /// </summary>
        public void BeforeEnvironmentDeactivate(object theEnvironment)
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.BeforeQuit event.
        /// </summary>
        public void BeforeQuit()
        {
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEApplicationEvents.BeforeWindowDeactivate event.
        /// </summary>
        public void BeforeWindowDeactivate(object theWindow)
        {
        }

        #endregion

        #region IConnectionPoint helpers

        /// <summary>
        /// Attaches specified events to this object.
        /// </summary>
        private void HookEvents(object eventSource, Guid eventGuid)
        {
            IConnectionPointContainer container = null;
            IConnectionPoint connectionPoint = null;
            int cookie = 0;

            container = (IConnectionPointContainer)eventSource;
            container.FindConnectionPoint(eventGuid, out connectionPoint);

            if (connectionPoint != null)
            {
                connectionPoint.Advise(this, out cookie);
                _connectionPointDictionary.Add(connectionPoint, cookie);
            }
        }

        /// <summary>
        /// Detaches specified events from this object.
        /// </summary>
        private void UnhookAllEvents()
        {
            Dictionary<IConnectionPoint, int>.Enumerator enumerator = _connectionPointDictionary.GetEnumerator();
            while (enumerator.MoveNext())
            {
                enumerator.Current.Key.Unadvise(enumerator.Current.Value);
            }

            _connectionPointDictionary.Clear();
        }

        #endregion

        #region regasm.exe

        /// <summary>
        /// Implementation of ComRegisterFunction.
        /// </summary>
        /// <remarks>
        /// This method gets called when regasm.exe is executed against the assembly.
        /// </remarks>
        [ComRegisterFunction]
        public static void Register(Type t)
        {
            var info = (AddInInfoAttribute)AddInInfoAttribute.GetCustomAttribute(t, typeof(AddInInfoAttribute));
            var environments = (AddInEnvironmentCategoryAttribute[])AddInEnvironmentCategoryAttribute.GetCustomAttributes(t, typeof(AddInEnvironmentCategoryAttribute));
            var autoConnect = (AddInAutoConnectAttribute)AddInAutoConnectAttribute.GetCustomAttribute(t, typeof(AddInAutoConnectAttribute));

            if (info == null) throw new System.Exception("Missing AddInInfoAttribute.");
            if ((environments == null) || (environments.Length == 0))
            {
                throw new System.Exception("Missing AddInEnvironmentCategoryAttribute.");
            }
            if (autoConnect == null) autoConnect = new AddInAutoConnectAttribute(true);

            string subkey = String.Format(@"CLSID\{0}", t.GUID.ToString("B"));
            using (RegistryKey baseKey = Registry.ClassesRoot.CreateSubKey(subkey))
            {
                subkey = String.Format(@"Implemented Categories\{0}", CategoryIDs.CATID_SolidEdgeAddIn);
                using (RegistryKey implementedCategoriesKey = baseKey.CreateSubKey(subkey))
                {
                }

                foreach (var environment in environments)
                {
                    subkey = String.Format(@"Environment Categories\{0}", environment.Guid.ToString("B"));
                    using (RegistryKey environmentCategoryKey = baseKey.CreateSubKey(subkey))
                    {
                    }
                }

                using (RegistryKey summaryKey = baseKey.CreateSubKey("Summary"))
                {
                    summaryKey.SetValue("409", info.Summary);
                }

                baseKey.SetValue("AutoConnect", autoConnect.AutoConnect ? 1 : 0);
                baseKey.SetValue("409", info.Title);
            }
        }

        /// <summary>
        /// Implementation of ComUnregisterFunction.
        /// </summary>
        /// <remarks>
        /// This method gets called when regasm.exe is executed against the assembly.
        /// </remarks>
        [ComUnregisterFunction]
        public static void Unregister(Type t)
        {
            string subkey = String.Format(@"CLSID\{0}", t.GUID.ToString("B"));
            Registry.ClassesRoot.DeleteSubKeyTree(subkey, false);
        }

        #endregion

        #region Properties

        public SolidEdgeFramework.Application Application { get { return _application; } }
        public SolidEdgeFramework.ISEAddInEx AddIn { get { return _addInEx; } }

        #endregion
    }
}
